src/nav2_tasks/include/nav2_tasks/TaskClient.hpp

```diff
@@ -68,48 +69,96 @@ class TaskClient
     typename ResultMsg::SharedPtr & result,
     std::chrono::milliseconds duration)
   {
-    std::mutex m;
-    std::unique_lock<std::mutex> lock(m);
-
-    std::cv_status timeoutStatus =
-      cv_.wait_for(lock, std::chrono::milliseconds(duration));
-
-    if (timeoutStatus == std::cv_status::timeout) {
+    // Wait for a status message to come in
+    std::unique_lock<std::mutex> lock(statusMutex_);
+    if (!cvStatus_.wait_for(lock, std::chrono::milliseconds(duration),
+      [&] {return statusReceived_ == true;}))
+    {
       return RUNNING;
     }
 
-    // TODO(orduno): possible race condition between receiving status message and actual data
-    //               for now implemented a method using a flag, but might want to review further
-    if (taskSucceeded_ && receivedResult_) {
-      result = result_;
-      receivedResult_ = false;
-      return SUCCEEDED;
+    // We've got a status message, indicating that the server task has finished (succeeded,
+    // failed, or canceled)
+    switch (statusMsg_->result) {
+      // If the task has failed or has been canceled, no result message is forthcoming and we
+      // can propagate the status code, using the TaskStatus type rather than the message-level
+      // implementation type
+      case nav2_tasks::msg::TaskStatus::FAILED:
+      case nav2_tasks::msg::TaskStatus::CANCELED:
+        return static_cast<TaskStatus>(statusMsg_->result);
+
+      case nav2_tasks::msg::TaskStatus::SUCCEEDED:
+        {
+          {
+            std::lock_guard<std::mutex> lock(resultMutex_);
+            if (resultReceived_) {
+              result = resultMsg_;
+              resultReceived_ = false;
+              return SUCCEEDED;
+            }
+          }
+
+          // The result message may have come *after* the status message, so let's wait for it
+          std::unique_lock<std::mutex> lock(resultMutex_);
+          if (cvResult_.wait_for(lock, std::chrono::milliseconds(100),
+            [&] {return resultReceived_ == true;}))
+          {
+            return SUCCEEDED;
```

Comment:
I think you need to set result = resultMsg_ and resultReceived_ = false in this block as well.